### PR TITLE
cyclicSort intro complete

### PR DESCRIPTION
**Pull request by xiaqian1999**

### CODE REVIEW

File: xiaqianZhang/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+# was assigned a unique number from 1 to ‘n’ based on their creation sequence. 
+# This means that the object with sequence number ‘3’ was created just before the object with sequence number ‘4’.
+
+# Write a function to sort the objects in-place on their creation sequence number in O(n) and without any extra space. 
+# For simplicity, let’s assume we are passed an integer array containing only the sequence numbers, 
+# though each number is actually an object.
+
+# input: unsorted list of object that is in incrementing order by 1, start from 1, ex: 1, 2, 3,...6
+# output: sorted list
+# edges case: if the len of the list <2: return the list
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.py:13]: Love that you wrote an edge case, but don't forget to write a test case to match as well!

File: xiaqianZhang/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+# was assigned a unique number from 1 to ‘n’ based on their creation sequence. 
+# This means that the object with sequence number ‘3’ was created just before the object with sequence number ‘4’.
+
+# Write a function to sort the objects in-place on their creation sequence number in O(n) and without any extra space. 
+# For simplicity, let’s assume we are passed an integer array containing only the sequence numbers, 
+# though each number is actually an object.
+
+#Input: unsorted list that contains distinct integers from 1 to n
+#Output: return sorted list
+# edge case: if the len(arr) < 2: example if only contains only 1 integers, return the arr since we don't need to sorted
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.py:48]: Loved the edge case, but don't forget to write a test case to match it!

File: xiaqianZhang/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+# initialize a currPtr position at index 0
+# while we looping through the nums, we want to check if each nums[i] is position at i-1
+# if no: having a swapPtr that is = to num[i-1], then swap the variable
+# if yes: i++
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.py:52]: Much better explanation!

File: xiaqianZhang/assignments/cyclicSort/lc268/lc268.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+#   return originalSum - currentSum
+
+
+#Solution with cyclicSort
+# We are given an array containing ‘n’ distinct numbers taken from the range 0 to ‘n’. 
+# Since the array has only ‘n’ numbers out of the total ‘n+1’ numbers, find the missing number.
+
+def find_missing_number(nums):
+
+
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/lc268/lc268.py:39]: Missing the coded solution. If you can not come up with the solution on your own, feel free to look at the solution provided and reverse-engineer it.

File: xiaqianZhang/assignments/cyclicSort/lc268/lc268.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+#   originalSum = 0
+#   currentSum = 0
+
+#   for each in range(len(nums)+1):
+#     originalSum+=each
+
+#   for each in range(len(nums)):
+#     currentSum+=nums[each]
+  
+#   return originalSum - currentSum
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/lc268/lc268.py:30]: Love this alternative solution!

File: xiaqianZhang/assignments/cyclicSort/lc268/lc268.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+# given n numbers, and the original length of the array is n+1
+# increment a space in the given array with variable x
+# while loop that swap the numbers if they are not on the right position, currentPtr = swapPtr, index 0 = num 0
+# ignore the x during the swap, and it will gradually end up in some position while the other integers are located at the correct position
+# return the index of where x located, that will be the missing integer
+
+# Solution without using the cyclicSort
+# original Sum = 0-4
+#currentSum = the numbers given from the input array
+#subtract the difference, that will be the missing number
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/lc268/lc268.py:18]: Great psuedo-code and attempt!

File: xiaqianZhang/assignments/cyclicSort/lc268/lc268.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
+# Problem Statement #
+
+# We are given an array containing ‘n’ distinct numbers taken from the range 0 to ‘n’. 
+# Since the array has only ‘n’ numbers out of the total ‘n+1’ numbers, find the missing number.
+
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/cyclicSort/lc268/lc268.py:5]: Missing edge cases

File: xiaqianZhang/assignments/mergeIntervals/lc56/lc56.py [hunk 8]
```diff
@@ -32,6 +32,19 @@
...
+# output: return the list of intervals without overlapping
+# edge case: if the len(intervals) < 2: we don't really need to do anything, just return the intervals
+# First taken the input intervals, we sort the intervals of its start where intervals looks like [start, end]
+# initialize start = intervals[0].start, end = intervals[0].end
+# loop through the entire list of intervals
+#   then initialize interval = intervals[i], while we loop through it, we want to compare intervals[i].end with intervals[i+1].start
+#   if intervals[i+1].start <= intervals[i].end ex: [1, 3] [2, 4] 2<3, it is overlapping
+#     update the end by taking the maximum from the intervals[i].end (end) with intervals[i+1].end (interval.end)
+#   else: merge the intervals and update the start and end
+
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/mergeIntervals/lc56/lc56.py:47]: Great psuedocode and edge case!

File: xiaqianZhang/assignments/mergeIntervals/lc56/lc56.py [hunk 9]
```diff
@@ -103,6 +141,11 @@ def main():
     i.print_interval()
   print()
 
+  print("Merged intervals: ", end='')
+  for i in merge([Interval(1, 4)]):
+    i.print_interval()
+  print()
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/mergeIntervals/lc56/lc56.py:147]: Love that you added a test case to match your edge case!

File: xiaqianZhang/assignments/slidingwindow/lc209/lc209.py [hunk 10]
```diff
@@ -12,19 +12,54 @@
...
+# Edge case: array is empty: return 0
+
+# initialize the windowStart = 0, and the currentSum = 0, constant update the currentSum to see if it's closer to the targetSum
+# initialize the shortestLength = math.inf, we want to initialize as large as possible
+# loop through each value in the array:
+#  update the currentSum += arr[each]
+#  if currentSum >= targetSum:
+#     update the shortestLength = min(shortestLength, currentIndex - windowStart + 1) 
+#   currentSum -= arr[each]
+#   windowStart+=1
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/slidingwindow/lc209/lc209.py:44]: Good psuedocode and high level approach.

File: xiaqianZhang/assignments/slidingwindow/lc53/lc53.py [hunk 11]
```diff
@@ -13,26 +14,89 @@
...
+
+# initialize the windowStart = 0, currentSum = 0, finalSum = 0
+# while windowStart < len(arr)-1: 
+# currentSum += arr[windowStart]
+#   if the windowStart > k:
+#       update the finalSum by taking the maximum among the currentSum and the finalSum
+#       currentSum -= arr[windowStart]
+#       windoeStart +=1
+
+#edge case: if the array is empty or the size of k is 0, return 0
```
Comment (knrodriguez) [inline xiaqianZhang/assignments/slidingwindow/lc53/lc53.py:80]: Great psuedocode and approach! Sometimes writing the approach in the whole sentences may clarify some concepts for you or for the interviewer.

**Review by knrodriguez — APPROVED** [review]

**Comment by knrodriguez** [comment]
I love that you are going back and revisiting problems to keep those concepts fresh! Keep it up!
